src/pubsub/ua_pubsub_connection.c

```diff
@@ -298,4 +303,234 @@ UA_PubSubConnection_regist(UA_Server *server, UA_NodeId *connectionIdentifier,
+}
+
+static UA_StatusCode
+setPubSubState_error(UA_Server *server,
+                                    UA_PubSubConnection *connection,
+                                    UA_StatusCode cause) {
+    UA_WriterGroup *wg;
+    UA_ReaderGroup *rg;
+    switch (connection->state){
+        case UA_PUBSUBSTATE_DISABLED:
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            break;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+            break;
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            LIST_FOREACH(wg, &connection->writerGroups, listEntry){
+                UA_WriterGroup_setPubSubState(server, wg, UA_PUBSUBSTATE_ERROR,
+                                                cause);
+            }
+            LIST_FOREACH(rg, &connection->readerGroups, listEntry){
+                UA_ReaderGroup_setPubSubState(server, rg, UA_PUBSUBSTATE_ERROR,
+                                                cause);
+            }
+
+            break;
+        case UA_PUBSUBSTATE_ERROR:
+            break;
+        default:
+            UA_LOG_WARNING_CONNECTION(&server->config.logger, connection,
+                            "Received unknown PubSub state!");
+    }
+    connection->state = UA_PUBSUBSTATE_ERROR;
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+UA_PubSubConnection_setPubSubState(UA_Server *server, UA_PubSubConnection *connection,
+                              UA_PubSubState state, UA_StatusCode cause) {
+    UA_LOCK_ASSERT(&server->serviceMutex, 1);
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    UA_PubSubState oldState = connection->state;
+
+    switch(state) {
+        case UA_PUBSUBSTATE_DISABLED:
+            ret = setPubSubState_disable(server, connection, cause);
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            ret = setPubSubState_paused(server, connection, cause);
+            break;
+        case UA_PUBSUBSTATE_PREOPERATIONAL:
+            ret = setPubSubState_preoperational(server, connection, cause);
+            break;
+        case UA_PUBSUBSTATE_OPERATIONAL:
+            ret = setPubSubState_operational(server, connection, cause);
+            break;
+        case UA_PUBSUBSTATE_ERROR: 
+            ret = setPubSubState_error(server, connection, cause);
+            break;
+        default:
+            UA_LOG_WARNING_CONNECTION(&server->config.logger, connection,
+                                       "Received unknown PubSub state!");
+    }
+    if(state != oldState) {
+        /* inform application about state change */
+        UA_ServerConfig *pConfig = &server->config;
+        if(pConfig->pubSubConfig.stateChangeCallback != 0) {
+            pConfig->pubSubConfig.
+                stateChangeCallback(server, &connection->identifier, state, cause);
+        }
+    }
+    return ret;
+}
+
+UA_StatusCode
+UA_Server_enablePubSubConnection(UA_Server *server,
+                                    const UA_NodeId connectionIdent) {
+    //UA_LOCK(&server->serviceMutex);
```

Comment:
You need to take the locks coming from the public API.
Create an "inner version" of the methods that assumes that locks are already taken.
If you need to call it also from within other server methods.